raft/rafttest: TestBasicProgress failed

raft/rafttest.TestBasicProgress [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16779862?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16779862?buildTab=artifacts#/) on release-24.2 @ [35538e048969d6d7937d6128dfd20ecdb4eb79d3](https://github.com/cockroachdb/cockroach/commits/35538e048969d6d7937d6128dfd20ecdb4eb79d3):


```
I240906 14:07:18.538678 294 (gostd) node.go:100  [-] 2  raft.2: stop
I240906 14:07:18.538695 296 (gostd) node.go:100  [-] 3  raft.3: stop
I240906 14:07:18.538708 298 (gostd) node.go:100  [-] 4  raft.4: stop
I240906 14:07:18.538722 300 (gostd) node.go:100  [-] 5  raft.5: stop
raft2024/09/06 14:07:18 INFO: 1 switched to configuration voters=()
raft2024/09/06 14:07:18 INFO: 1 became follower at term 0
raft2024/09/06 14:07:18 INFO: newRaft 1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/09/06 14:07:18 INFO: 1 became follower at term 1
raft2024/09/06 14:07:18 INFO: 1 switched to configuration voters=(1)
raft2024/09/06 14:07:18 INFO: 1 switched to configuration voters=(1 2)
raft2024/09/06 14:07:18 INFO: 1 switched to configuration voters=(1 2 3)
raft2024/09/06 14:07:18 INFO: 1 switched to configuration voters=(1 2 3 4)
raft2024/09/06 14:07:18 INFO: 1 switched to configuration voters=(1 2 3 4 5)
raft2024/09/06 14:07:18 INFO: 2 switched to configuration voters=()
raft2024/09/06 14:07:18 INFO: 2 became follower at term 0
raft2024/09/06 14:07:18 INFO: newRaft 2 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/09/06 14:07:18 INFO: 2 became follower at term 1
raft2024/09/06 14:07:18 INFO: 2 switched to configuration voters=(1)
raft2024/09/06 14:07:18 INFO: 2 switched to configuration voters=(1 2)
raft2024/09/06 14:07:18 INFO: 2 switched to configuration voters=(1 2 3)
raft2024/09/06 14:07:18 INFO: 2 switched to configuration voters=(1 2 3 4)
raft2024/09/06 14:07:18 INFO: 2 switched to configuration voters=(1 2 3 4 5)
raft2024/09/06 14:07:18 INFO: 3 switched to configuration voters=()
raft2024/09/06 14:07:18 INFO: 3 became follower at term 0
raft2024/09/06 14:07:18 INFO: newRaft 3 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/09/06 14:07:18 INFO: 3 became follower at term 1
raft2024/09/06 14:07:18 INFO: 3 switched to configuration voters=(1)
raft2024/09/06 14:07:18 INFO: 3 switched to configuration voters=(1 2)
raft2024/09/06 14:07:18 INFO: 3 switched to configuration voters=(1 2 3)
raft2024/09/06 14:07:18 INFO: 3 switched to configuration voters=(1 2 3 4)
raft2024/09/06 14:07:18 INFO: 3 switched to configuration voters=(1 2 3 4 5)
raft2024/09/06 14:07:18 INFO: 4 switched to configuration voters=()
raft2024/09/06 14:07:18 INFO: 4 became follower at term 0
raft2024/09/06 14:07:18 INFO: newRaft 4 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/09/06 14:07:18 INFO: 4 became follower at term 1
raft2024/09/06 14:07:18 INFO: 4 switched to configuration voters=(1)
raft2024/09/06 14:07:18 INFO: 4 switched to configuration voters=(1 2)
raft2024/09/06 14:07:18 INFO: 4 switched to configuration voters=(1 2 3)
raft2024/09/06 14:07:18 INFO: 4 switched to configuration voters=(1 2 3 4)
raft2024/09/06 14:07:18 INFO: 4 switched to configuration voters=(1 2 3 4 5)
raft2024/09/06 14:07:18 INFO: 5 switched to configuration voters=()
raft2024/09/06 14:07:18 INFO: 5 became follower at term 0
raft2024/09/06 14:07:18 INFO: newRaft 5 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/09/06 14:07:18 INFO: 5 became follower at term 1
raft2024/09/06 14:07:18 INFO: 5 switched to configuration voters=(1)
raft2024/09/06 14:07:18 INFO: 5 switched to configuration voters=(1 2)
raft2024/09/06 14:07:18 INFO: 5 switched to configuration voters=(1 2 3)
raft2024/09/06 14:07:18 INFO: 5 switched to configuration voters=(1 2 3 4)
raft2024/09/06 14:07:18 INFO: 5 switched to configuration voters=(1 2 3 4 5)
--- FAIL: TestBasicProgress (5.21s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBasicProgress.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-41954

--- a/pkg/raft/rafttest/BUILD.bazel
+++ b/pkg/raft/rafttest/BUILD.bazel
@@ -52,6 +52,5 @@ go_test(
     deps = [
         "//pkg/raft",
         "//pkg/raft/raftpb",
-        "//pkg/testutils/skip",
     ],
 )

--- a/pkg/raft/rafttest/node_test.go
+++ b/pkg/raft/rafttest/node_test.go
@@ -24,7 +24,6 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/raft"
 	pb "github.com/cockroachdb/cockroach/pkg/raft/raftpb"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 )
 
 func TestBasicProgress(t *testing.T) {
@@ -54,9 +53,6 @@ func TestBasicProgress(t *testing.T) {
 }
 
 func TestRestart(t *testing.T) {
-	// TODO(pav-kv): de-flake it. See https://github.com/etcd-io/raft/issues/181.
-	skip.UnderStress(t, "the test is flaky")
-
 	peers := []raft.Peer{{ID: 1, Context: nil}, {ID: 2, Context: nil}, {ID: 3, Context: nil}, {ID: 4, Context: nil}, {ID: 5, Context: nil}}
 	nt := newRaftNetwork(1, 2, 3, 4, 5)
 
@@ -136,15 +132,14 @@ func TestPause(t *testing.T) {
 }
 
 func waitLeader(ns []*node) int {
-	var l map[pb.PeerID]struct{}
-	var lindex int
-
+	l := make(map[pb.PeerID]struct{})
 	for {
-		l = make(map[pb.PeerID]struct{})
+		clear(l)
+		lindex := -1
 
 		for i, n := range ns {
 			lead := n.Status().SoftState.Lead
-			if lead != 0 {
+			if lead != raft.None {
 				l[lead] = struct{}{}
 				if n.id == lead {
 					lindex = i
@@ -152,7 +147,7 @@ func waitLeader(ns []*node) int {
 			}
 		}
 
-		if len(l) == 1 {
+		if len(l) == 1 && lindex != -1 {
 			return lindex
 		}
 	}

